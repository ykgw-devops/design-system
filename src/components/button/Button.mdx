---
name: Button
menu: Components
---

import { Playground, Props } from 'docz'
import Button from './Button'
import Group from './Group'
import Icon from '../icon/Icon'
import { Popup } from '../popup/Popup'
import Segment from '../segment/Segment'

# Button

## Basic usage

<Playground>
  <Button>Primary</Button>

  <Button>
    <Icon name='add_shopping_cart' /> Add to cart
  </Button>

  <Button outline kind='danger'>
    <Icon name='stop' /> Stop recording
  </Button>
</Playground>

## As link

<Playground>
  <Button as='a' href='https://google.com/' target='_blank'>Google <Icon name='launch' /></Button>
</Playground>

## Kind

<Playground>
  <Button kind="primary">Primary</Button>
  <Button kind="secondary">Secondary</Button>
  <Button kind="warning">Warning</Button>
  <Button kind="danger">Danger</Button>
  <Button kind="success">Success</Button>
</Playground>

## Pill

<Playground>
  <Button pill kind="danger">Red Pill</Button>
  <Button pill kind="primary">Blue Pill</Button>
</Playground>

## Outline

<Playground>
  <Button kind="primary" outline>Primary</Button>
  <Button kind="secondary" outline>Secondary</Button>
  <Button kind="warning" outline>Warning</Button>
  <Button kind="danger" outline>Danger</Button>
  <Button kind="success" outline>Success</Button>
</Playground>

## Disabled

<Playground>
  <Button disabled>Disabled</Button>

  <Button pill outline kind="warning" disabled>Disabled Warning Button</Button>
</Playground>

## Size

<Playground>
  <Button size='tiny'>Tiny</Button>

  <Button size='small'>Small</Button>

  <Button size='medium'>Medium</Button>

  <Button size='large'>Large</Button>
</Playground>

## Loading

<Playground>
  <Button loading>Hold on tight...</Button>
  <Button kind='success' loading>Hold on tight...</Button>
  <Button kind='danger' loading>Hold on tight...</Button>
  <br />
  <Button disabled loading>Hold on tight...</Button>
  <Button disabled kind='success' loading>Hold on tight...</Button>
  <Button disabled kind='danger' loading>Hold on tight...</Button>
  <br />
  <Button outline loading>Hold on tight...</Button>
  <Button outline kind='success' loading>Hold on tight...</Button>
  <Button outline kind='danger' loading>Hold on tight...</Button>
  <br />
  <Button disabled outline loading>Hold on tight...</Button>
  <Button disabled outline kind='success' loading>Hold on tight...</Button>
  <Button disabled outline kind='danger' loading>Hold on tight...</Button>
</Playground>

## Properties

<Props of={Button}></Props>

# Group

<Playground>
  Base:
  <Button.Group>
    <Button>
      <Icon name='filter_1' /> Option #1
    </Button>
    <Button outline>Option #2</Button>
    <Button>
      <Icon name='filter_3' /> Option #3
    </Button>
  </Button.Group>

  <br />

  Dropdown Button:
  <Button.Group>
    <Button>
      <Icon name='filter_1' /> Option #1
    </Button>
    <Popup
      placement='bottom-end'
      appendTo={() => document.body}
      content={
        <Segment>
          Some Dropdown Content Here
        </Segment>
      }
    >
      <Button>
        <Icon name='keyboard_arrow_down' />
      </Button>
    </Popup>
  </Button.Group>

  <br />

  <Button.Group>
    <Button kind='secondary'>
      <Icon name='filter_1' /> Option #1
    </Button>
    <Button kind='secondary'>Option #2</Button>
    <Button outline kind='secondary'>
      <Icon name='filter_3' /> Option #3
    </Button>
  </Button.Group>

  <br />

  Compact width:
  <Button.Group compact>
    <Button>Number #1</Button>
    <Button kind='danger'>Number #2</Button>
    <Button kind='success'>Number #3</Button>
  </Button.Group>

  <br />

  Fluid:
  <Button.Group fluid>
    <Button>Text</Button>
    <Button kind='danger'>Text</Button>
    <Button kind='success'>How much wood would a woodchuck chuck if a woodchuck could chuck wood?</Button>
  </Button.Group>

  <br />

  Equal width:
  <Button.Group equalWidth>
    <Button>Number #1</Button>
    <Button kind='danger'>Number #2</Button>
    <Button kind='success'>How much wood would a woodchuck chuck if a woodchuck could chuck wood?</Button>
  </Button.Group>
</Playground>

## Properties

<Props of={Group}></Props>
