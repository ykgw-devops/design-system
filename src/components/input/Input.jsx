import styled from '@emotion/styled'
import PropTypes from 'prop-types'

import Group from './Group'
import { input, fluid as fluidStyle, basic as basicStyle, adornments } from './Input.styles'

const Input = styled.input`
  ${input};
  ${props => props.fluid && fluidStyle};
  ${props => props.basic && basicStyle};
`

const Adornment = styled.span(adornments)

Input.Adornment = Adornment
Input.Group = Group

Input.propTypes = {
  onChange: PropTypes.func,
  adornments: PropTypes.array,
  fluid: PropTypes.bool
}

Input.defaultProps = {
  onChange: () => undefined,
  adornments: [],
  fluid: false
}

export default Input
