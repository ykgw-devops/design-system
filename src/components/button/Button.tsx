import React, { ButtonHTMLAttributes } from 'react'
import styled from '@emotion/styled'
import isPropValid from '@emotion/is-prop-valid'
import { setDisplayName } from 'recompose'

import Group from './Group'
import Loader from '../loader/Loader'

import { base, kind, outline, pill, disabled, size, colorFromProps } from './Button.styles'

type ButtonKind = 'primary' | 'secondary' | 'warning' | 'danger' | 'success'
type ButtonSize = 'tiny' | 'small' | 'medium' | 'large'

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  outline?: boolean;
  pill?: boolean;
  disabled?: boolean;
  loading?: boolean;
  kind?: ButtonKind;
  size?: ButtonSize;
}

const ValidButton = styled('button', { shouldForwardProp: isPropValid })

const Button = ValidButton<IButtonProps>`
  ${base};
  ${props => kind(props)};
  ${props => size(props)};
  ${props => props.outline && outline(props)};
  ${props => props.pill && pill};
  ${props => props.disabled && disabled(props)};
`

const EnhancedButton = React.forwardRef((props: IButtonProps, ref: React.Ref<any>) => {
  const { loading, children, ...rest } = props

  return (
    <Button {...rest} ref={ref}>
      {loading && (
        <>
          <Loader
            color={colorFromProps(props)}
            size={14}
            thickness={2}
          />
          &nbsp;
        </>
      )}
      {children}
    </Button>
  )
})

EnhancedButton.Group = setDisplayName('Button.Group')(Group)

EnhancedButton.defaultProps = {
  type: 'button',
  kind: 'primary',
  size: 'medium',
  pill: false,
  outline: false,
  disabled: false,
  loading: false
}

export default EnhancedButton
