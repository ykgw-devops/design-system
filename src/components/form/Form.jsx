import React from 'react'
import setDisplayName from 'recompose/setDisplayName'
import PropTypes from 'prop-types'

import { base } from './Form.styles.jsx'
import Field from './Field'
import Input from '../input/Input'

const Form = props => {
  return (
    <form {...props} css={base}>
      {props.children}
    </form>
  )
}

const noop = (event) => {
  event.preventDefault()
  event.stopPropagation()
}

Form.propTypes = {
  onSubmit: PropTypes.func
}

Form.defaultProps = {
  onSubmit: noop
}

Form.Input = setDisplayName('Form.Input')(Input)
Form.Field = setDisplayName('Form.Field')(Field)

export default Form
