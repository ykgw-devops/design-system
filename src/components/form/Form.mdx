---
name: Form
menu: Components
---

import { Playground, Props } from 'docz'
import Button from '../button/Button'
import Form from './Form'


# Form

## Basic usage

<Playground>
  <Form style={{ width: 250 }}>
    <Form.Field>
      <label>Email</label>
      <Form.Input name='email' placeholder='example@domain.com' type='email' autoComplete='login-email' />
    </Form.Field>
    <Form.Field>
      <label>Password</label>
      <Form.Input name='password' placeholder='*****' type='password' autoComplete='login-password' />
    </Form.Field>
    <div>
      <Button outline kind='secondary'>Cancel</Button>
      <Button>Submit</Button>
    </div>
  </Form>
</Playground>

## Properties

<Props of={Form}></Props>
