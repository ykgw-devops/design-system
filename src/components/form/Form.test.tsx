import '@testing-library/jest-dom'
import React from 'react'
import { render } from '@testing-library/react'
import serializer from 'jest-emotion'

import Input from '../input/Input'
import Button from '../button/Button'
import Form from './Form'

expect.addSnapshotSerializer(serializer)

test('button spacing in group', () => {
  const { asFragment } = render(
    <Form>
      <Input.Group>
        <Button>Hello, World!</Button>
      </Input.Group>
    </Form>
  )

  expect(asFragment()).toMatchSnapshot()
})
