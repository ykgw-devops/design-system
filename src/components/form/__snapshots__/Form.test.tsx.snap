// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button spacing in group 1`] = `
<DocumentFragment>
  .emotion-4 {
  display: block;
}

.emotion-4 input,
.emotion-4 label {
  display: block;
}

.emotion-4 label {
  color: #59595B;
}

.emotion-4:focus {
  color: red;
}

.emotion-4 button {
  margin-right: 8px;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  border-radius: 0.333333rem;
  background: white;
}

.emotion-2 > * {
  box-sizing: border-box;
  font-size: 1rem;
  padding: 0 0.85rem;
  line-height: 1rem;
  height: 2.4rem;
  display: inline-block;
  vertical-align: middle;
}

.emotion-2 input {
  border: solid 1px #E8E8E8;
  margin: 0;
}

.emotion-2 input + button,
.emotion-2 button + button {
  border-left: none;
}

.emotion-2 button + input {
  border-left: none;
}

form .emotion-2 button {
  margin-right: 0;
}

.emotion-2 button {
  border-radius: 0;
  background: #f5f5f5;
  color: inherit;
  border: solid 1px #E8E8E8;
}

.emotion-2 button:hover {
  background: #ececec;
}

.emotion-2 > *:first-child {
  border-top-left-radius: 0.333333rem;
  border-bottom-left-radius: 0.333333rem;
}

.emotion-2 > *:last-child {
  border-top-right-radius: 0.333333rem;
  border-bottom-right-radius: 0.333333rem;
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: pre;
  text-align: center;
  background-color: #1E90FF;
  border-radius: 0.333333rem;
  border: solid 1px #1b81e5;
  color: #fff;
  cursor: pointer;
  font-family: "Fira Sans",system-ui,sans-serif;
  font-weight: 400;
  padding: 0.65em 1em;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 80ms ease-in;
  transition: background 80ms ease-in;
  vertical-align: middle;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #1E90FF;
  color: white;
  border-color: #1b81e5;
  font-size: 0.9rem;
  line-height: 0.9rem;
}

.emotion-0:hover {
  background-color: #1b81e5;
}

.emotion-0:hover {
  background-color: #1b81e5;
}

<form
    class="emotion-4"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <button
        class="emotion-0 emotion-1"
        kind="primary"
        type="button"
      >
        Hello, World!
      </button>
    </div>
  </form>
</DocumentFragment>
`;
